### PR TITLE
Improve MaskedSoftmax by oneDNN

### PR DESCRIPTION
**Pull request by bgawrych**
## Description ##
Utilize few oneDNN primitives to mask input and execute softmax

Accuracy comparision on GluonNLP models (not affected) [C6i.16xlarge]:
image

Performance comaprision (samples/s) [C6i.16xlarge]: 
image

BERT Base can archive up to 60% more samples/s with this change

### CODE REVIEW

**Comment by bgawrych** [comment]
@mxnet-bot run ci [centos-cpu, unix-cpu]

**Comment by bgawrych** [comment]
@mxnet-bot run ci [windows-gpu]

File: src/operator/nn/dnnl/dnnl_masked_softmax.cc [hunk 1]
```diff
@@ -0,0 +1,195 @@
...
+
+void DNNLMaskedSoftmaxFwd::Execute(const Tensors& tensors,
+                                   const std::vector<OpReqType>& req,
+                                   const MaskedSoftmaxParam& param,
+                                   bool is_train) const {
+  using namespace mxnet::op;
+  using namespace mxnet::op::mxnet_op;
+  using namespace dnnl;
+  /*
+   1. out = input * [(mask - 1) * inf]
```
Comment (bartekkuncer) [inline src/operator/nn/dnnl/dnnl_masked_softmax.cc:136]: Maybe add a title to this e.g. Three steps of masked softmax.

File: src/operator/nn/dnnl/dnnl_softmax.cc [hunk 2]
```diff
@@ -195,6 +218,24 @@ DNNLSoftmaxBwd& DNNLSoftmaxBwd::GetCached(const SoftmaxParam& param, const Tenso
...
+  const int axis            = CheckAxis(param.axis, tensors.out.shape().ndim());
+  const auto out_grad_mem   = tensors.out_grad.GetDNNLData();
+  const auto out_mem        = tensors.out.GetDNNLData();
+  const auto softmax_fwd_pd = DNNLSoftmaxFwd::GetSoftmaxFwdPd(*out_mem, axis, /*is_train*/ true);
+
+  softmax_bwd_pd = std::make_shared<softmax_bwd_pd_t>(
+      GetSoftmaxBwdPd(*out_grad_mem, *out_mem, axis, softmax_fwd_pd));
+  softmax_bwd = std::make_shared<softmax_bwd_t>(*softmax_bwd_pd);
+
+  if (temperature != 1.0f) {  // avoid dividing by 1
```
Comment (bartekkuncer) [inline src/operator/nn/dnnl/dnnl_softmax.cc:232]: Should we not check somewhere if temperature == 0?

File: src/operator/nn/masked_softmax.cc [hunk 3]
```diff
@@ -0,0 +1,141 @@
...
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+/*!
+ * \file masked_softmax.cc
+ */
+
+#include "./softmax-inl.h"
+#include "../tensor/elemwise_unary_op.h"
```
Comment (bartekkuncer) [inline src/operator/nn/masked_softmax.cc:25]: Please avoid dots in includes.

File: src/operator/nn/dnnl/dnnl_softmax.cc (hunk 2 above)
Comment (bgawrych) [reply src/operator/nn/dnnl/dnnl_softmax.cc:232]: There is no restriction in norml version so I belive we shouldn't check it either

File: src/operator/nn/dnnl/dnnl_masked_softmax.cc [hunk 4]
```diff
@@ -0,0 +1,196 @@
...
+      dnnl::memory(mask_desc, engine, reinterpret_cast<int8_t*>(mask.data().dptr<bool>()));
+  stream->RegisterPrimArgs(this->primitives->minusone,
+                           {
+                               {DNNL_ARG_SRC, mask_mem},
+                               {DNNL_ARG_DST, *mask_minusone_mem},
+                           });
+
+  // 1. B) out = out * inf
+  const memory::desc converted_mask_desc = this->primitives->mask_input_pd.src1_desc();
+  dnnl::memory* convertedmask_mem        = TmpMemMgr::Get()->Alloc(converted_mask_desc);
```
Comment (bartekkuncer) [inline src/operator/nn/dnnl/dnnl_masked_softmax.cc:160]:
```suggestion
  dnnl::memory* converted_mask_mem       = TmpMemMgr::Get()->Alloc(converted_mask_desc);
```

**Review by bartekkuncer — APPROVED** [review]
LGTM

**Comment by bgawrych** [comment]
@mxnet-bot run ci [centos-gpu]

**Review by mozga-intel — APPROVED** [review]

**Pull request merged**